plotly/graph_reference.py

```diff
@@ -0,0 +1,355 @@
+                if isinstance(val, dict) and val.get('_isSubplotObj'):
+                    subplot_names.add(key)
+
+    response['valid_names'] = valid_names
+    response['deprecated_names'] = deprecated_names
+    response['subplot_names'] = subplot_names
+
+    return response
+
+
+def _is_valid_sub_path(path, parent_paths):
+    """
+    Check if a sub path is valid given an iterable of parent paths.
+
+    :param (tuple[str]) path: The path that may be a sub path.
+    :param (list[tuple]) parent_paths: The known parent paths.
+    :return: (bool)
+
+    Examples:
+
+        * ('a', 'b', 'c') is a valid subpath of ('a', )
+        * ('a', 'd') is not a valid subpath of ('b', )
+        * ('a', ) is not a valid subpath of ('a', 'b')
+        * ('anything',) is a valid subpath of ()
+
+    """
+    if not parent_paths:
+        return True
+    for parent_path in parent_paths:
+        if path[:len(parent_path)] == parent_path:
+            return True
+    return False
+
+
+def _get_hr_name(object_name):
+    """Get human readable object name from reference ('hrName')."""
+    if object_name in ARRAYS:
+        return object_name  # TODO: how to handle hr name on array?
+
+    object_dict = OBJECTS[object_name]
+    meta_paths = object_dict['meta_paths']
+    for meta_path in meta_paths:
+        meta_dict = utils.get_by_path(GRAPH_REFERENCE, meta_path)
+        if meta_dict.get('_isLinkedToArray'):
+            return object_name  # TODO: how to handle hr name on array?
+        if 'hrName' in meta_dict:
+            return meta_dict['hrName']
+
+    return object_name  # default is to keep the original name
+
+
+def _get_objects():
+    """
+    Create a reorganization of graph reference which organizes by object name.
+
+    Each object can have *many* different definitions in the graph reference.
+    These possibilities get narrowed down when we have contextual information
+    about parent objects. For instance, Marker in Scatter has a different
+    definition than Marker in Pie. However, we need Marker, Scatter, and Pie
+    to exist on their own as well.
+
+    Each value has the form:
+    {
+        'meta_paths': [],
+        'attribute_paths': [],
+        'additional_attributes': {}
+    }
+
+    * meta_paths describes the top-most path where this object is defined
+    * attribute_paths describes all the locations where attributes exist
+    * additional_attributes can be used to hard-code (patch) the plot schema
+
+    :return: (dict)
+
+    """
+    objects = {}
+    for node, path in utils.node_generator(GRAPH_REFERENCE):
+        if any([key in path for key in GRAPH_REFERENCE['defs']['metaKeys']]):
+            continue  # objects don't exist under nested meta keys
+
+        # note that arrays are *not* stored in objects! they're arrays!
+        if node.get('role') == 'object':
+            object_name = path[-1]
+            if node.get('_isLinkedToArray'):
+                object_name = object_name[:-1]
```

Comment:
Ha, yeah. This is pretty fragile though. @etpinard can we add a test that ensures that `_isLinkedToArray` objects in the schema end in an `s` or something? Else, can we enforce that such arrays define some meta information about the items they have? `_itemName` or something? I don't have a good sense of how this works out will arrays that have many items. For example, would `data` (not that it's explicitely defined) use `"_itemName": "scatter"` as a default?